File: Programmers/코딩테스트_고득점_Kit/Greedy/level3_단속카메라.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42884?language=python3
# 정렬해 접근하는 방식까지는 생각해냈으나, 자꾸 오답이 출력되어 결국 솔루션 봄
# 생각보다 알고리즘 자체는 매우 간단한 편


def solution(routes):
    routes.sort(key=lambda x: x[1])  # 도착 지점 기준으로 정렬
    cnt_camera = 1
    current_arrival = routes[0][1]  # 초기값 지정
    for start, end in routes[1:]:
        if start > current_arrival:  # 새로운 루트의 시작지점이 현재 도착지점의 뒤에 있는 경우
            cnt_camera += 1  # 카메라 개수 추가
            current_arrival = end  # 도착 지점 갱신
    return cnt_camera

File: Programmers/코딩테스트_고득점_Kit/Greedy/level3_단속카메라.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/42884?language=cpp
#include <algorithm>
#include <vector>
using namespace std;

bool cmp(vector<int> a, vector<int> b) {
  return a[1] < b[1];  // 도착점 기준 오름차순 정렬
}

int solution(vector<vector<int>> routes) {
  sort(routes.begin(), routes.end(), cmp);
  int cnt_camera = 1;
  int current_arrival = routes[0][1];  // 초기값 지정
  for (int i = 1; i < routes.size(); i++) {
    if (routes[i][0] > current_arrival) {
      cnt_camera += 1;
      current_arrival = routes[i][1];
    }
  }
  return cnt_camera;
}